src/java/org/jivesoftware/openfire/session/DomainPair.java

```diff
@@ -0,0 +1,38 @@
+package org.jivesoftware.openfire.session;
+
+/**
+ * Holds a (possibly authenticated) domain pair.
+ */
+public class DomainPair {
+    private final String local;
+    private final String remote;
+
+    public DomainPair(String local, String remote) {
+        this.local = local;
+        this.remote = remote;
+    }
+
+    public int hashCode() {
```

Comment:
Even though it is *probably* alright, the implementation of `hashCode` and `equals` aught to be consistent. By far the easiest way to guarantee this is to let your IDE generate these methods for you. In Intellij: [ALT-INSERT (select "generate equals() and hashCode()")](https://www.jetbrains.com/help/idea/generate-equals-and-hashcode-wizard.html). Generated code also lets you stop worry about null references and the like.

This is of particular importance here, as you're storing instances of `DomainPair` in a HashSet (in `LocalOutgoingServerSession`)

Reply:
Done.